api/resolvers/paidAction.js

```diff
@@ -50,28 +50,44 @@ export default {
     }
   },
   Mutation: {
-    retryPaidAction: async (parent, { invoiceId }, { models, me, lnd }) => {
+    retryPaidAction: async (parent, { invoiceId, newAttempt }, { models, me, lnd }) => {
       if (!me) {
         throw new Error('You must be logged in')
       }
 
-      const invoice = await models.invoice.findUnique({ where: { id: invoiceId, userId: me.id } })
+      // make sure only one client at a time can retry by immediately transitioning to an intermediate state
+      const [invoice] = await models.$queryRaw`
+          UPDATE "Invoice"
+          SET "actionState" = 'RETRY_PENDING'
+          WHERE id in (
+            SELECT id FROM "Invoice"
+            WHERE id = ${invoiceId} AND "userId" = ${me.id} AND "actionState" = 'FAILED'
+            FOR UPDATE
+          )
```

Comment:
Why use a set query here? Wouldn't this suffice:

```sql
          UPDATE "Invoice"
          SET "actionState" = 'RETRY_PENDING'
          WHERE id = ${invoiceId} AND "userId" = ${me.id} AND "actionState" = 'FAILED'
          RETURNING *
```

`id` is a unique column so there'd only ever be one.

Reply:
Mhh, this `SELECT FOR UPDATE` was probably a remainder of the previous query which locked multiple invoices at once and I used `SELECT FOR UPDATE` because I wanted to use `SKIP LOCKED`.

I don't use `SKIP LOCKED` anymore since none of the locks are acquired in long-running transactions. So the update queries never try long to acquire the lock. There would only ever be a delay while another update is trying to immediately acquire the lock or fail but that should be insignificant.